packages/dd-trace/src/span_link_processor.js

```diff
@@ -0,0 +1,113 @@
+const traceContextInjector = new TextMapPropagator({ tracePropagationStyle: { inject: 'tracecontext' } })
+
+const ALLOWED = ['string', 'number', 'boolean']
+const ALLOWED_SPAN_LINKS_KEY = [ 'trace_id', 'span_id', 'attributes',
+  'dropped_attributes_count', 'tracestate', 'trace_id_high', 'flags']
+
+function sanitizeAttributes (formattedLink, attributes) {
+  let _attributesString = '{'
+
+  const addArrayorScalarAttributes = (key, maybeArray) => {
+    if (Array.isArray(maybeArray)) {
+      for (const subkey in maybeArray) {
+        addArrayorScalarAttributes(`${key}.${subkey}`, maybeArray[subkey])
+      }
+    } else {
+      const maybeScalar = maybeArray
+      if (ALLOWED.includes(typeof maybeScalar)) {
+        formattedLink.attributesCount++
+        if (_attributesString.length === 1) { // no attributes yet
+          _attributesString += `"${key}":"${maybeScalar}"}`
+        } else {
+          _attributesString = _attributesString.slice(0, -1) + `,"${key}":"${maybeScalar}"}`
+        }
+      } else {
+        log.warn(`Dropping span link attribute.`)
+        formattedLink.dropped_attributes_count++
+      }
+    }
+  }
+
+  Object
+    .entries(attributes)
+    .forEach(entry => {
+      const [key, value] = entry
+      addArrayorScalarAttributes(key, value)
+    })
+
+  return `${_attributesString}`
+}
+
+function getTraceFlags (spanContext, linkContext, formattedLink) {
+  // _tracestate only set when w3c trace flags are given
+  if (spanContext?._tracestate) {
+    formattedLink.flags = spanContext._sampling.priority > 0 ? 1 : 0
+  }
+
+  // take the linkContext's tracestate over the spanContext's if it exists
+  if (linkContext?._tracestate) {
+    formattedLink.flags = linkContext._sampling.priority > 0 ? 1 : 0
+  }
+}
+
+function formatLink (spanContext, linkContext, attributes = {}) {
+  const formattedLink = {
+    trace_id: linkContext._traceId.toString(),
+    span_id: linkContext._spanId.toString(),
+    dropped_attributes_count: 0,
+    attributesCount: 0
+  }
+  formattedLink.attributes = sanitizeAttributes(formattedLink, attributes)
+
+  if (!linkContext._traceId.equals(spanContext._traceId)) return formattedLink
+
+  getTraceFlags(spanContext, linkContext, formattedLink)
+
+  formattedLink.trace_id_high = linkContext?._trace.tags['_dd.p.tid'] || spanContext?._trace.tags['_dd.p.tid']
+  formattedLink.tracestate = linkContext?._tracestate || spanContext?._tracestate
+
+  // inject extracted Datadog HTTP headers into local tracestate
+  // indicated by _trace.origin
+  if (!formattedLink.tracestate && spanContext?._trace?.origin) {
+    const extractedTracestate = {}
+    traceContextInjector.inject(spanContext, extractedTracestate)
+    formattedLink.tracestate = extractedTracestate.tracestate
+  }
+
+  return {
+    ...formattedLink,
+    trace_id: formattedLink.trace_id_high
+      ? formattedLink.trace_id_high + formattedLink.trace_id : formattedLink.trace_id,
+    tracestate: formattedLink.tracestate?.toString()
```

Comment:
Similarly, why set it to anything other than a string if a string will be used here?